src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +624,231 @@ internal SafeFileHandle OpenPort(string portName)
+            long totalBytesRead = 0;
+            // last value of totalBytesRead + BytesToRead
+            long lastTotalBytesAvailable = BytesToRead;
+            while (IsOpen && !eofReceived && !_ioLoopFinished)
+            {
+                // PollEvents will wait if both flags are false
+                Interop.Sys.PollEvents events = PollEvents(1,
+                                                           pollReadEvents: _readQueue.Count > 0,
+                                                           pollWriteEvents: _writeQueue.Count > 0);
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLNVAL) ||
+                    events.HasFlag(Interop.Sys.PollEvents.POLLERR))
+                {
+                    // bad descriptor or some other error we can't handle
+                    break;
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLIN))
+                {
+                    int bytesRead = DoIORequest(_readQueue, ProcessRead);
+                    totalBytesRead += bytesRead;
+                }
+
+                // check if there is any new data (either already read or in the driver input)
+                // this event is private and handled inside of SerialPort
+                // which then throttles it with the threshold
+                long totalBytesAvailable = totalBytesRead + BytesToRead;
+                if (totalBytesAvailable > lastTotalBytesAvailable)
+                {
+                    lastTotalBytesAvailable = totalBytesAvailable;
+
+                    // We need new task so that this thread doesn't get deadlocked when someone calls
+                    // Read from within the event
+
+                    Task.Factory.StartNew(s => {
+                            var thisRef = (SerialStream)s;
+                            thisRef.DataReceived(thisRef, new SerialDataReceivedEventArgs(SerialData.Chars));
+                        },
+                        this,
+                        CancellationToken.None,
+                        TaskCreationOptions.DenyChildAttach,
+                        TaskScheduler.Default);
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLOUT))
+                {
+                    DoIORequest(_writeQueue, ProcessWrite);
+                }
+            }
+        }
+
+        private static CancellationToken GetCancellationTokenFromTimeout(int timeoutMs)
+        {
+            return timeoutMs == SerialPort.InfiniteTimeout ?
+                CancellationToken.None :
+                (new CancellationTokenSource(Math.Max(timeoutMs, TimeoutResolution))).Token;
+        }
+
+        class SerialStreamAsyncResult
+        {
+            public Memory<byte> Buffer { get; private set; }
+            public bool IsCompleted => _tcs.Task.IsCompleted;
+            public Task<int> UnderlyingTask => _tcs.Task;
+
+            private TaskCompletionSource<int> _tcs = new TaskCompletionSource<int>();
```

Comment:
We should create this with TaskCreationOptions.RunContinuationsAsynchronously as the argument constructor.  When someone calls tcs.SetResult/Exception, by default any continuations off of the associated task will be allowed to run synchronously, which means we'll be invoking them from that dedicated loop thread on which we only want to run our code.  Specifying that flag forces continuations to be asynchronous from SetResult/Exception.